_posts/2017-10-06-rust-blog-series-2.md

```diff
@@ -0,0 +1,1098 @@
+        .expect("DATABASE_URL must be set");
+    PgConnection::establish(&database_url)
+        .expect(&format!("Error connecting to {}", database_url))
+}
+```
+
+This is a great start.
+Let's review the code here so we can understand the pros and cons of this approach.
+The first two lines import the libs we need.
+We need Diesel for our ORM/Query building and [dotenv] to pull out our database url we've stored in the `.env` file
+from [Part I] of this series.
+
+[dotenv]: https://github.com/purpliminal/rust-dotenv
+[Part I]: 2017-09-15-rust-blog-series-1.md // FIXME BAD LINK
+
+The next four `use` statments bring some modules into scope from those libraries.
+`diesel::prelude::*;` imports a whole lot of wonderful parts of the diesel api,
+which allow us build sequel queries out of diesel methods.
+
+Turn your attention to the `establish_connection()` function.
+
+```rust
+pub fn establish_connection() -> PgConnection {
+    dotenv().ok(); // Grab ENV vars from `.env`
+
+    // Pull value from `DATABASE_URL` ENV var
+    // `expect()` is used because env::var returns a Result
+    // https://doc.rust-lang.org/std/env/fn.var.html
+    let database_url = env::var("DATABASE_URL")
+        .expect("DATABASE_URL must be set");
+
+    // Establishes a connection to the DB
+    // again, `expect()` is used because a ConnectionResult is returned.
+    // https://docs.diesel.rs/diesel/connection/trait.Connection.html#tymethod.establish
+    PgConnection::establish(&database_url)
+        .expect(&format!("Error connecting to {}", database_url))
+}
+```
+
+`establish_connection` returns a [`PgConnection`], which will let us establish a connection to our Postgres database.
+If we were to use Sqlite or Mysql,
+then we would use the appropriate diesel imports for those backends and NOT use `pg::PgConnection`.
+
+Remember that `.env` file from Part I of the series?
+Well, this is the first important piece of code where we need it.
+`dotenv().ok();` loads the environment variables from our `.env` file, which should be in the current directory
+or any of the parent directories. Documentation for `dotenv` can be found [here](https://github.com/purpliminal/rust-dotenv).
+
+The `database_url` grabs the database url string from the environment variables we just loaded and unwrapped via the `expect()` method.
+We can then call the [`establish()`] method provided by `PgConnection` and pass in the `database_url`, which will grant us a connection to
+our Postgres database!
+
+[Diesel Getting Started]: http://diesel.rs/guides/getting-started/
+[`PgConnection`]: https://docs.diesel.rs/diesel/pg/struct.PgConnection.html
+[`establish()`]: https://docs.diesel.rs/diesel/connection/trait.Connection.html#tymethod.establish
+
+
+## Connection Pools
+
+The previous code works well in isolation, but also poses some scalability problems.
+I think I [remember reading somewhere] that establishing a database connection is expensive.
```

Comment:
It's actually much more important when using an ORM, as there's generally a prepared statement cache that is persisted between requests. It's even more important in Diesel because our prepared statement cache lets us skip the query builder entirely. [Relevant internal docs](https://github.com/diesel-rs/diesel/blob/d30821a6625a574b77ab9aaaeedf7283676d3946/diesel/src/connection/statement_cache.rs#L1-L91)

Reply:
Also connection pooling is useful even if it's just a TCP connection (which is why browsers re-use connections for HTTP requests). Even just the cost of a TCP handshake is worth pooling.